projects/vdk-core/src/vdk/api/plugin/plugin_input.py

```diff
@@ -214,6 +277,25 @@ def ingest_payload(
         """
         pass
 
+    def pre_ingest_process(self, payload: List[dict]) -> List[dict]:
+        """
+        Do some processing on the ingestion payload before passing it to the
+        actual ingestion.
+
+        For example, if we want to ensure that all values in the payload are
+        string before the actual ingestion happens, we could have a plugin,
+        implementing this method, and doing some payload processing:
+
+        .. code-block:: python
+            def pre_ingest_process(self, payload: List[dict]) -> List[dict]:
+                # Ensure all values in the payload are strings
+                return [{k: str(v) for (k,v) in i.items()} for i in payload]
+
+        :param payload: List[dict] the ingestion payload to be processed.
+        :return: List[dict], the processed payload.
```

Comment:
Could we also recieve and return, along with the payload, a dict containing metadata? For example:
```
NewType("IngestionMetadata", Dict)

def pre_ingest_process(self, payload: List[dict], metadata: IngestionMetadata) -> List[dict], Optional?[IngestionMetadata]:
```
This is common to the Optional[IngestionResult] we return after (successful) ingestion. 


The use case for this ask is, we need to be able to send metadata about ingestion start time for example. Or length of the payload in case it has been non-successfully ingested. For example, a pre-processor would populate:
```
'send_start_time': datetime.datetime.utcnow().isoformat(),
'send_end_time': None,
'http_status': -1
```
Then, a post-processor could overwrite to populate send_end_time.

That metadata dictionary, needs to be available for each pre-processing, ingestion and and post-processing step - to be read or modified. Note: all of those steps happen sequentially.

Reply:
I've added the additional interface feature, though am not sure to what extent would it be used.